File: backend/src/main/java/com/workflow/tramites/Tramite.java
package com.workflow.tramites;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.CompoundIndexes;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "tramites")
@CompoundIndexes({
    @CompoundIndex(name = "idx_cliente_estado", def = "{'cliente_id': 1, 'estado': 1}"),
    @CompoundIndex(name = "idx_etapa_rol", def = "{'etapa_actual.responsable_rol_nombre': 1, 'estado': 1}"),
    @CompoundIndex(name = "idx_politica_estado", def = "{'politica_id': 1, 'estado': 1}"),
    @CompoundIndex(name = "idx_etapa_rol_asignado", def = "{'etapa_actual.responsable_rol_nombre': 1, 'asignado_a_id': 1}")
})
public class Tramite {

    public enum EstadoTramite {
        INICIADO, EN_PROCESO, SIN_ASIGNAR, COMPLETADO, RECHAZADO, CANCELADO, DEVUELTO, ESCALADO, EN_APELACION
    }

    public enum EstadoApelacion { PENDIENTE, EN_REVISION, APROBADO, DENEGADO }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Apelacion {
        private boolean activa;

        @Field("fecha_inicio")
        private LocalDateTime fechaInicio;

        @Field("fecha_limite")
        private LocalDateTime fechaLimite;

        @Field("motivo_original")
        private String motivoOriginal;

        @Field("documentos_originales")
        private List<com.workflow.files.FileReference> documentosOriginales;

        @Field("documentos_apelatoria")
        private List<com.workflow.files.FileReference> documentosApelatoria;

        @Field("justificacion_cliente")
        private String justificacionCliente;

        private EstadoApelacion estado;
    }

    // -----------------------------------------------------------------------
    // Clases embebidas
    // -----------------------------------------------------------------------

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class EtapaActual {
        @Field("actividad_bpmn_id")
        private String actividadBpmnId;

        @Field("nombre")
        private String nombre;

        @Field("responsable_rol_nombre")
        private String responsableRolNombre;

        @Field("formulario_id")
        private String formularioId;

        @Field("area")
        private String area;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class HistorialEntry {
        @Field("actividad_bpmn_id")
        private String actividadBpmnId;

        @Field("actividad_nombre")
        private String actividadNombre;

        @Field("responsable_id")
        private String responsableId;

        @Field("responsable_nombre")
        private String responsableNombre;

        @Field("accion")
        private String accion;

        @Field("timestamp")
        private LocalDateTime timestamp;

        @Field("observaciones")
        private String observaciones;

        @Field("responsable_cargo")
        private String responsableCargo;

        @Field("documentos_adjuntos")
        private List<com.workflow.files.FileReference> documentosAdjuntos;
    }

    // -----------------------------------------------------------------------
    // Campos del documento
    // -----------------------------------------------------------------------

    @Id
    private String id;

    @Field("politica_id")
    @Indexed
    private String politicaId;

    @Field("politica_nombre")
    private String politicaNombre;

    @Field("politica_version")
    private Integer politicaVersion;

    @Field("cliente_id")
    @Indexed
    private String clienteId;

    @Field("cliente_nombre")
    private String clienteNombre;

    @Field("asignado_a_id")
    @Indexed
    private String asignadoAId;

    @Field("asignado_a_nombre")
    private String asignadoANombre;

    @Field("estado")
    @Builder.Default
    private EstadoTramite estado = EstadoTramite.INICIADO;

    @Field("etapa_actual")
    private EtapaActual etapaActual;

    @Field("historial")
    @Builder.Default
    private List<HistorialEntry> historial = new ArrayList<>();

    @Field("creado_en")
    private LocalDateTime creadoEn;

    @Field("actualizado_en")
    private LocalDateTime actualizadoEn;

    @Field("fecha_vencimiento_etapa")
    private LocalDateTime fechaVencimientoEtapa;

    @Field("apelacion")
    private Apelacion apelacion;
}

File: mobile/lib/data/models/tramite.dart
class FileRef {
  final String? fileId;
  final String? nombre;
  final String? url;

  const FileRef({this.fileId, this.nombre, this.url});

  factory FileRef.fromJson(Map<String, dynamic> j) => FileRef(
        fileId: j['fileId'] as String?,
        nombre: j['nombre'] as String?,
        url: j['url'] as String?,
      );
}

class EtapaActualDTO {
  final String? actividadBpmnId;
  final String? nombre;
  final String? responsableRolNombre;
  final String? formularioId;
  final String? area;

  const EtapaActualDTO({
    this.actividadBpmnId,
    this.nombre,
    this.responsableRolNombre,
    this.formularioId,
    this.area,
  });

  factory EtapaActualDTO.fromJson(Map<String, dynamic> j) => EtapaActualDTO(
        actividadBpmnId: j['actividadBpmnId'] as String?,
        nombre: j['nombre'] as String?,
        responsableRolNombre: j['responsableRolNombre'] as String?,
        formularioId: j['formularioId'] as String?,
        area: j['area'] as String?,
      );
}

class HistorialEntryDTO {
  final String? actividadBpmnId;
  final String? actividadNombre;
  final String? responsableId;
  final String? responsableNombre;
  final String? accion;
  final DateTime? timestamp;
  final String? observaciones;
  final String? responsableCargo;
  final List<FileRef> documentosAdjuntos;
  final Map<String, dynamic> datos;

  const HistorialEntryDTO({
    this.actividadBpmnId,
    this.actividadNombre,
    this.responsableId,
    this.responsableNombre,
    this.accion,
    this.timestamp,
    this.observaciones,
    this.responsableCargo,
    this.documentosAdjuntos = const [],
    this.datos = const {},
  });

  factory HistorialEntryDTO.fromJson(Map<String, dynamic> j) {
    final rawDocs = j['documentosAdjuntos'] as List<dynamic>? ?? [];
    final rawDatos = j['datos'];

    return HistorialEntryDTO(
      actividadBpmnId: j['actividadBpmnId'] as String?,
      actividadNombre: j['actividadNombre'] as String?,
      responsableId: j['responsableId'] as String?,
      responsableNombre: j['responsableNombre'] as String?,
      accion: j['accion'] as String?,
      timestamp: j['timestamp'] != null
          ? DateTime.tryParse(j['timestamp'] as String)
          : null,
      observaciones: j['observaciones'] as String?,
      responsableCargo: j['responsableCargo'] as String?,
      documentosAdjuntos: rawDocs
          .map((e) => FileRef.fromJson(e as Map<String, dynamic>))
          .toList(),
      datos: rawDatos is Map<String, dynamic> ? rawDatos : const {},
    );
  }
}

class ApelacionDTO {
  final bool activa;
  final DateTime? fechaInicio;
  final DateTime? fechaLimite;
  final String? motivoOriginal;
  final List<FileRef> documentosOriginales;
  final List<FileRef> documentosApelatoria;
  final String? justificacionCliente;
  final String? estado;

  const ApelacionDTO({
    this.activa = false,
    this.fechaInicio,
    this.fechaLimite,
    this.motivoOriginal,
    this.documentosOriginales = const [],
    this.documentosApelatoria = const [],
    this.justificacionCliente,
    this.estado,
  });

  /// True when an appeal is active and hasn't been resolved yet.
  bool get pendiente => activa && (estado == null || estado == 'PENDIENTE');

  factory ApelacionDTO.fromJson(Map<String, dynamic> j) {
    final rawOrig = j['documentosOriginales'] as List<dynamic>? ?? [];
    final rawApel = j['documentosApelatoria'] as List<dynamic>? ?? [];

    return ApelacionDTO(
      activa: (j['activa'] as bool?) ?? false,
      fechaInicio: j['fechaInicio'] != null
          ? DateTime.tryParse(j['fechaInicio'] as String)
          : null,
      fechaLimite: j['fechaLimite'] != null
          ? DateTime.tryParse(j['fechaLimite'] as String)
          : null,
      motivoOriginal: j['motivoOriginal'] as String?,
      documentosOriginales: rawOrig
          .map((e) => FileRef.fromJson(e as Map<String, dynamic>))
          .toList(),
      documentosApelatoria: rawApel
          .map((e) => FileRef.fromJson(e as Map<String, dynamic>))
          .toList(),
      justificacionCliente: j['justificacionCliente'] as String?,
      estado: j['estado'] as String?,
    );
  }
}

class TramiteResponse {
  final String id;
  final String? politicaId;
  final String? politicaNombre;
  final int? politicaVersion;
  final String? clienteId;
  final String? clienteNombre;
  final String? asignadoAId;
  final String? asignadoANombre;
  final String estado;
  final EtapaActualDTO? etapaActual;
  final List<HistorialEntryDTO> historial;
  final DateTime? creadoEn;
  final DateTime? actualizadoEn;
  final DateTime? fechaVencimientoEtapa;
  final ApelacionDTO? apelacion;

  const TramiteResponse({
    required this.id,
    this.politicaId,
    this.politicaNombre,
    this.politicaVersion,
    this.clienteId,
    this.clienteNombre,
    this.asignadoAId,
    this.asignadoANombre,
    required this.estado,
    this.etapaActual,
    this.historial = const [],
    this.creadoEn,
    this.actualizadoEn,
    this.fechaVencimientoEtapa,
    this.apelacion,
  });

  factory TramiteResponse.fromJson(Map<String, dynamic> j) {
    final rawHistorial = j['historial'] as List<dynamic>? ?? [];

    EtapaActualDTO? etapa;
    final rawEtapa = j['etapaActual'];
    if (rawEtapa is Map<String, dynamic>) {
      etapa = EtapaActualDTO.fromJson(rawEtapa);
    }

    ApelacionDTO? apelacion;
    final rawApelacion = j['apelacion'];
    if (rawApelacion is Map<String, dynamic>) {
      apelacion = ApelacionDTO.fromJson(rawApelacion);
    }

    return TramiteResponse(
      id: (j['id'] as String?) ?? '',
      politicaId: j['politicaId'] as String?,
      politicaNombre: j['politicaNombre'] as String?,
      politicaVersion: j['politicaVersion'] as int?,
      clienteId: j['clienteId'] as String?,
      clienteNombre: j['clienteNombre'] as String?,
      asignadoAId: j['asignadoAId'] as String?,
      asignadoANombre: j['asignadoANombre'] as String?,
      estado: (j['estado'] as String?) ?? 'INICIADO',
      etapaActual: etapa,
      historial: rawHistorial
          .map((e) => HistorialEntryDTO.fromJson(e as Map<String, dynamic>))
          .toList(),
      creadoEn: j['creadoEn'] != null
          ? DateTime.tryParse(j['creadoEn'] as String)
          : null,
      actualizadoEn: j['actualizadoEn'] != null
          ? DateTime.tryParse(j['actualizadoEn'] as String)
          : null,
      fechaVencimientoEtapa: j['fechaVencimientoEtapa'] != null
          ? DateTime.tryParse(j['fechaVencimientoEtapa'] as String)
          : null,
      apelacion: apelacion,
    );
  }
}
